scripts/train_linker.py

```diff
@@ -7,60 +7,103 @@
 import argparse
 from tqdm import tqdm
 import json
+import numpy as np
+from collections import defaultdict
 
 def read_file(filename, limit):
+        data = []
         x = []
         y = []
         with open(filename) as f:
             for line in tqdm(f, total=limit):
                 d = json.loads(line)
+                data.append(d)
                 x.append(Linker.featurizer(d))
                 y.append(d['label'])
                 if len(x) >= limit:
                     break
-        return x, y
+        return data, x, y
 
-def main(data_path: str):
+def entity_level_eval(data, all_features, classifier):
+    k_list = [1, 3, 5, 10]
+    classifier_correct_predictions = defaultdict(int)
+    classifier_total_predictions = defaultdict(int)
+    candidate_features = []
+    gold_index = -1
+    for i, d in tqdm(enumerate(data)):
+        candidate_index = d['candidate_index']
+        if candidate_index == 0:
+            if len(candidate_features) > 0:
+                if gold_index != -1:  # compute "normalized" recall
+                    scores = classifier.predict_proba(candidate_features)
+                    sorted_ids = np.argsort(-scores[:, 1], kind='mergesort')
+                    for k in k_list:
+                        if gold_index in sorted_ids[:k]:
+                            classifier_correct_predictions[k] += 1
+                        classifier_total_predictions[k] += 1
+            candidate_features = []
+            gold_index = -1
+
+        assert candidate_index == len(candidate_features)
+        if d['label'] == 1:
```

Comment:
Why does the label of the current piece of data effect the processing of the next piece of data? I do not understand what's going on here....